pkg/webhook/cspc_operations.go

```diff
@@ -509,3 +468,213 @@ func GetNewBDFromRaidGroups(newRG, oldRG *cstor.RaidGroup) map[string]string {
+		mapOldBlockDevices[bd.BlockDeviceName] = true
+	}
+	for _, bd := range newRg.CStorPoolInstanceBlockDevices {
+		if !mapOldBlockDevices[bd.BlockDeviceName] {
+			bds = append(bds, bd.BlockDeviceName)
+		}
+	}
+	return bds
+}
+
+// getBDsFromRaidGroups return list of blockdevices from list of raid groups
+func getBDsFromRaidGroups(rgs []cstor.RaidGroup) []string {
+	bds := []string{}
+	for _, rg := range rgs {
+		for _, bd := range rg.CStorPoolInstanceBlockDevices {
+			bds = append(bds, bd.BlockDeviceName)
+		}
+	}
+	return bds
+}
+
+// getExpandedRaidGroups returns the raid groups only if pool expansion is done
+// by the users
+func getExpandedRaidGroups(
+	raidGroups []cstor.RaidGroup, oldRaidGroups []cstor.RaidGroup) []cstor.RaidGroup {
+	expandedRgs := []cstor.RaidGroup{}
+	for _, newRg := range raidGroups {
+		isRaidGroupExist := false
+		for _, oldRg := range oldRaidGroups {
+			if IsRaidGroupCommon(oldRg, newRg) {
+				isRaidGroupExist = true
+				break
+			}
+		}
+		// If no common blockdevices exists in raid groups then newRg is
+		// added for expansion
+		if !isRaidGroupExist {
+			expandedRgs = append(expandedRgs, newRg)
+		}
+	}
+	return expandedRgs
+}
+
+// validateNewBDs returns nil if new BDs are valid for expansion or else it will
+// return error if occurs
+func (pOps *PoolOperations) validateNewBDs(newBDs []string, cspc *cstor.CStorPoolCluster) error {
+	for _, bd := range newBDs {
+		bdc, err := pOps.GetBDCOfBD(bd)
+		if err != nil {
+			return errors.Wrapf(err, "failed to get claim of block device %s", bd)
+		}
+		// The incoming BD should be a valid entry if
+		// 1. The BD does not have a BDC.
+		// 2. The BD has a BDC with the current CSPC label and there is no successor of this BD
+		//    present in the CSPC.
+		if !pOps.IsBDValid(bd, bdc, cspc) {
+			return errors.Errorf("can not use blockdevice %s validation failed", bd)
+		}
+	}
+	return nil
+}
+
+// validatePoolExpansion will validate only expanded raid groups or new blockdevices(
+// in stripe only block devices are added). Following are the validations:
+// 1. New blockdevice shouldn't be claimed by any other CSPC (or) third party.
+// 2. New blockdevice shouldn't be the replacing blockdevice.
+func (pOps *PoolOperations) validatePoolExpansion(
+	newPoolSpec *cstor.PoolSpec, commonRaidGroups map[string]*raidGroups) error {
+	var bds []string
+	for rgType, rgs := range commonRaidGroups {
+		if rgs.rgType == string(cstor.PoolStriped) {
+			bds = append(bds, getNewBDsFromStripeSpec(rgs.oldRaidGroups[0],
+				rgs.newRaidGroups[0])...)
+		} else {
+			if rgType == "data" {
+				newRgs := getExpandedRaidGroups(newPoolSpec.DataRaidGroups, rgs.oldRaidGroups)
+				bds = getBDsFromRaidGroups(newRgs)
+			}
+			if rgType == "writeCache" {
+				newRgs := getExpandedRaidGroups(newPoolSpec.WriteCacheRaidGroups, rgs.oldRaidGroups)
+				bds = getBDsFromRaidGroups(newRgs)
+			}
+		}
+	}
+	err := pOps.validateNewBDs(bds, pOps.NewCSPC)
```

Comment:
In validateNewBDs function we are calling `IsBDValid`. IsBDValid function is expecting oldCSPC but we passed NewCSPC. If we look from here it makes sense to pass oldCSPC since we already have new blockDevices in our hand.
```go
func (pOps *PoolOperations) IsBDValid(bd string, bdc *openebsapis.BlockDeviceClaim, oldcspc *cstor.CStorPoolCluster) bool {
}
```